### PR TITLE
Fix buggy behaviour when addressing derived types

### PR DESCRIPTION
**Pull request by gathogojr**
### Issues

*This pull request fixes issue #2277.*

### Description

Based on the [OData specification](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_AddressingDerivedTypes), the following two requests should be applicable when restricting an entity to a derived type instance

http://host/service/Customers/Model.VipCustomer(1) - key appearing after the cast
http://host/service/Customers(1)/Model.VipCustomer - key appearing before the cast (right after entity set)

In addition, if the customer with key 1 is not a VipCustomer, the result should be a 404 Not Found

This PR addresses that. Also added missing tests for these scenarios.

**Note:**
In the [**EntitySetRoutingConvention**](https://github.com/OData/WebApi/blob/ade333a0d063c0b18787623c1fd5651d0a497ed1/src/Microsoft.AspNet.OData.Shared/Routing/Conventions/EntitySetRoutingConvention.cs#L66), the _~/entityset/cast_ path template is handled by either **Get[EntitySet]From[EntityType]** (e.g. `GetCustomersFromVipCustomer()`) or **GetFrom[EntityType]** (e.g. `GetFromVipCustomer()`). However, in regard to the [**EntityRoutingConventation**](https://github.com/OData/WebApi/blob/ade333a0d063c0b18787623c1fd5651d0a497ed1/src/Microsoft.AspNet.OData.Shared/Routing/Conventions/EntityRoutingConvention.cs#L20), the _~/entityset/key/cast_ path template is handled by **Get[EntityType]** (e.g. `GetVipCustomer(int key)`) and `Get(int key)`. This is actually what our documentation says on [routing conventions](https://docs.microsoft.com/en-us/odata/webapi/built-in-routing-conventions). For this reason, where the relevant action method is not implemented, a 404 Not Found is returned and may create the impression that there's a bug/feature gap.

Q. Would there be any added benefit in having the _~/entityset/key/cast_ path template be handled _**additionally**_ by `GetCustomersFromVipCustomer(int key)` and `GetFromVipCustomer(int key)` to mirror the equivalent methods for derived type entity set routing?

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Review by marabooy — APPROVED** [review]
LGTM

**Comment by paulodero** [comment]
> ### Issues
> _This pull request fixes issue #2277._
> 
> ### Description
> Based on the [OData specification](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_AddressingDerivedTypes), the following two requests should be applicable when restricting an entity to a derived type instance
> 
> http://host/service/Customers/Model.VipCustomer(1) - key appearing after the cast
> http://host/service/Customers(1)/Model.VipCustomer - key appearing before the cast (right after entity set)
> 
> In addition, if the customer with key 1 is not a VipCustomer, the result should be a 404 Not Found
> 
> This PR addresses that. Also added missing tests for these scenarios.
> 
> **Note:**
> In the [**EntitySetRoutingConvention**](https://github.com/OData/WebApi/blob/ade333a0d063c0b18787623c1fd5651d0a497ed1/src/Microsoft.AspNet.OData.Shared/Routing/Conventions/EntitySetRoutingConvention.cs#L66), the _~/entityset/cast_ path template is handled by either **Get[EntitySet]From[EntityType]** (e.g. `GetCustomersFromVipCustomer()`) or **GetFrom[EntityType]** (e.g. `GetFromVipCustomer()`). However, in regard to the [**EntityRoutingConventation**](https://github.com/OData/WebApi/blob/ade333a0d063c0b18787623c1fd5651d0a497ed1/src/Microsoft.AspNet.OData.Shared/Routing/Conventions/EntityRoutingConvention.cs#L20), the _~/entityset/key/cast_ path template is handled by **Get[EntityType]** (e.g. `GetVipCustomer(int key)`) and `Get(int key)`. This is actually what our documentation says on [routing conventions](https://docs.microsoft.com/en-us/odata/webapi/built-in-routing-conventions). For this reason, where the relevant action method is not implemented, a 404 Not Found is returned and may create the impression that there's a bug/feature gap.
> 
> Q. Would there be any added benefit in having the _~/entityset/key/cast_ path template be handled _**additionally**_ by `GetCustomersFromVipCustomer(int key)` and `GetFromVipCustomer(int key)` to mirror the equivalent methods for derived type entity set routing?
> 
> ### Checklist (Uncheck if it is not completed)
> * [x]  _Test cases added_
> * [ ]  _Build and test with one-click build and test script passed_
> 
> ### Additional work necessary
> _If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue._

@gathogojr from the documentation and your explanation looks like there is inconsistency in how _~/entityset/cast_ path template and _~/entityset/key/cast_ path template are handled. One of them could have adopted the other's style. To your question, I don't see any other added benefit apart from consistency which improves predictability. In my opinion if we add support for `GetCustomerFromVipCustomer(int key)` and `GetFromVipCustomer(int key)` then _~/entityset/cast_ should also be handled by **Get[EntitySet]** and **Get** as well which I guess now complicates the situation. 

For _~/entityset/key/cast_ I think you meant `GetCustomerFromVipCustomer(int key)` without an **s**

**Review by paulodero — APPROVED** [review]
:shipit:

**Review by odero — APPROVED** [review]

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DerivedTypes/DerivedTypesTests.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+            var response = await this.Client.SendAsync(request);
+            Assert.True(response.IsSuccessStatusCode);
+
+            var expectedContent = string.Format("{{\"@odata.context\":\"{0}/odata/$metadata" +
+                "#Customers/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer/$entity\"," +
+                "\"Id\":2,\"Name\":\"Customer 2\",\"LoyaltyCardNo\":\"9876543210\"}}", this.BaseAddress);
+
+            Assert.Equal(expectedContent, await response.Content.ReadAsStringAsync());
+        }
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DerivedTypes/DerivedTypesTests.cs:97]: Merge these two test cases using 

[Theory]
[InlineData("Customers/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer(2)")]
[InlineData("Customers(2)/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer")]

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DerivedTypes/DerivedTypesTests.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                "\"Orders\":[{{\"Id\":2,\"Amount\":230}},{{\"Id\":3,\"Amount\":150}}]}}]}}", this.BaseAddress);
+
+            Assert.Equal(expectedContent, await response.Content.ReadAsStringAsync());
+        }
+
+        [Fact]
+        public async Task RestrictEntityToDerivedTypeInstance_ThenExpandNavProperty()
+        {
+            // Key preceeds name of the derived type
+            string requestUri = this.BaseAddress + "/odata/Customers(2)/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer?$expand=Orders";
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DerivedTypes/DerivedTypesTests.cs:134]: > "/odata/Customers(2)/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer?$expand=Orders" [](start = 50, length = 94)

can it work for use $expand after the "/.../cast/key"? Maybe a test to cover it? 
Expand this case to using [Theory]

**Pull request merged**